src/common/buffer.cc

```diff
@@ -1501,39 +1502,50 @@ using namespace ceph;
   void buffer::list::append(char c)
   {
     // put what we can into the existing append_buffer.
-    unsigned gap = append_buffer.unused_tail_length();
+    unsigned gap = get_append_buffer_unused_tail_length();
     if (!gap) {
-      // make a new append_buffer!
-      append_buffer = raw_combined::create(CEPH_BUFFER_APPEND_SIZE, 0,
-					   get_mempool());
-      append_buffer.set_length(0);   // unused, so far.
+      // make a new buffer!
+      auto buf = ptr_node::create(
+	raw_combined::create(CEPH_BUFFER_APPEND_SIZE, 0, get_mempool()));
+      buf->set_length(0);   // unused, so far.
+      _buffers.push_back(*buf.release());
     }
-    append(append_buffer, append_buffer.append(c) - 1, 1);	// add segment to the list
+    _buffers.back().append(c);
+    _len++;
   }
 
   void buffer::list::append(const char *data, unsigned len)
   {
+    // hmm, maybe let's provide ::appends with guarantee the container is
+    // never empty?
+    if (_buffers.empty()) {
+      size_t need = round_up_to(len, sizeof(size_t)) + sizeof(raw_combined);
+      size_t alen = round_up_to(need, CEPH_BUFFER_ALLOC_UNIT) -
+	sizeof(raw_combined);
+      _buffers.emplace_back(raw_combined::create(alen, 0, get_mempool()));
+      _buffers.back().set_length(0);   // unused, so far.
+    }
+
     while (len > 0) {
-      // put what we can into the existing append_buffer.
-      unsigned gap = append_buffer.unused_tail_length();
-      if (gap > 0) {
-        if (gap > len) gap = len;
-    //cout << "append first char is " << data[0] << ", last char is " << data[len-1] << std::endl;
-        append_buffer.append(data, gap);
-        append(append_buffer, append_buffer.length() - gap, gap);	// add segment to the list
-        len -= gap;
-        data += gap;
+      ptr& last_one = _buffers.back();
+      unsigned gap = \
+	std::min(last_one.raw_nref() == 1 ? last_one.unused_tail_length() : 0, len);
```

Comment:
is this faster than wrapping is in an if (last_one.raw_nref() == 1) ?

Reply:
Yup, it's pointless. https://github.com/ceph/ceph/pull/25077/commits/b9e12f4a5482f058064911f3075484392ed7368d switches to:
```cpp
    const unsigned free_in_last = get_append_buffer_unused_tail_length();
    const unsigned first_round = std::min(len, free_in_last);
    if (first_round) {
      _buffers.back().append(data, first_round);
    }
```